core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -213,6 +222,77 @@ private[log] class LogCleanerManager(val logDirs: Seq[File],
+      val now = time.milliseconds
+      this.timeOfLastRun = now
+      val lastClean = allCleanerCheckpoints
+
+      val candidateLogs = inLock(lock) {
+        logs.filter {
+          case (_, log) => log.config.compact // match logs that are marked as compacted
+        }.filterNot {
+          case (topicPartition, log) =>
+            // skip any logs already in-progress and uncleanable partitions
+            inProgress.contains(topicPartition) || isUncleanablePartition(log, topicPartition)
+        }
+      }
+
+      val dirtyLogs = candidateLogs.map {
+        case (topicPartition, log) => // create a LogToClean instance for each
+          try {
+            val lastCleanOffset = lastClean.get(topicPartition)
+            val offsetsToClean = cleanableOffsets(log, lastCleanOffset, now)
+            // update checkpoint for logs with invalid checkpointed offsets
+            if (offsetsToClean.forceUpdateCheckpoint)
+              updateCheckpoints(log.parentDirFile, partitionToUpdateOrAdd = Option(topicPartition, offsetsToClean.firstDirtyOffset))
+            val compactionDelayMs = maxCompactionDelay(log, offsetsToClean.firstDirtyOffset, now)
+            preCleanStats.updateMaxCompactionDelay(compactionDelayMs)
+
+            LogToClean(topicPartition, log, offsetsToClean.firstDirtyOffset, offsetsToClean.firstUncleanableDirtyOffset, compactionDelayMs > 0)
+          } catch {
+            case e: Throwable => throw new LogCleaningException(log,
+              s"Failed to calculate log cleaning stats for partition $topicPartition", e)
+          }
+      }.filter(ltc => ltc.totalBytes > 0) // skip any empty logs
+
+      this.dirtiestLogCleanableRatio = if (dirtyLogs.nonEmpty) dirtyLogs.max.cleanableRatio else 0
+      // and must meet the minimum threshold for dirty byte ratio or have some bytes required to be compacted
+      var cleanableLogs = dirtyLogs.filter { ltc =>
+        (ltc.needCompactionNow && ltc.cleanableBytes > 0) || ltc.cleanableRatio > ltc.log.config.minCleanableRatio
+      }
+      inLock(lock) {
+      // the logs inside cleanableLogs may have changed when we are not holding the lock, e.g. they could have become
+      // 1. removed from the logs due to topic deletion
+      // 2. aborted (with the log's value in the inProgress map being LogCleaningAborted)
+      // 3. in progress in other cleaner threads
+      // 4. uncleanable due to LogCleaningExceptions
+      // We shouldn't proceed with logs in any of these condition
+      //
+      cleanableLogs = cleanableLogs.filter {
+        logToClean => logs.contains(logToClean.topicPartition)
+      }.filterNot {
+        logToClean =>
+          inProgress.contains(logToClean.topicPartition) || isUncleanablePartition(logToClean.log, logToClean.topicPartition)
+      }
+      if(cleanableLogs.isEmpty) {
+        None
+      } else {
+        preCleanStats.recordCleanablePartitions(cleanableLogs.size)
+        val filthiest = cleanableLogs.max
```

Comment:
Not an issue specifically for this PR, but it seems like there's a *huge* amount of work going on here to identify just one log, and virtually all of it is going to get repeated at least once if not several (or many?) times. I don't have a sense for how frequently this method is called, but if it's enough to show up as a hot spot, it's probably worth rethinking the approach altogether.

For example, what if, in addition to the single worst log, it also produces a map of the next 10 or 20 candidates, and then the top of the method is modified to check that map, throw out the same 4 categories as immediately above this line, and go with what remains until empty? That should be superfast, and while it means THE dirtiest log might not get cleaned up at the earliest possible moment, if the map isn't too big, the delay between "earliest possible moment" and when it actually happens won't be too large. (And the speedup factor from such caching would be close to the size of the map.)

Reply:
I agree, such improvements can be helpful. But I'll leave it as future work.